lib/wda/webdriveragent.js

```diff
@@ -249,6 +252,50 @@ class WebDriverAgent {
     return await this.xcodebuild.start();
   }
 
+  async startWithIDB () {
+    log.info('Will launch WDA with idb instead of xcodebuild since it is provided');
+    const {wdaBundleId, testBundleId} = await this.prepareWDA();
+    const env = {
+      USE_PORT: this.wdaRemotePort,
+      WDA_PRODUCT_BUNDLE_IDENTIFIER: this.updatedWDABundleId,
+    };
+    if (this.mjpegServerPort) {
+      env.MJPEG_SERVER_PORT = this.mjpegServerPort;
+    }
+
+    return await this.idb.runXCUITest(wdaBundleId, wdaBundleId, testBundleId, {env});
+  }
+
+  async parseBundleId (wdaBundlePath) {
+    const infoPlistPath = path.join(wdaBundlePath, 'Info.plist');
+    const infoPlist = await plist.parsePlist(await fs.readFile(infoPlistPath));
+    if (!infoPlist.CFBundleIdentifier) {
+      throw new Error(`Couldn't find bundle id in ${infoPlistPath}`);
+    }
+    return infoPlist.CFBundleIdentifier;
+  }
+
+  async prepareWDA () {
+    const wdaBundlePath = await this.fetchWDABundle();
+    const wdaBundleId = await this.parseBundleId(wdaBundlePath);
+    if (!await this.device.isAppInstalled(wdaBundleId)) {
+      await this.device.installApp(wdaBundlePath);
+    }
+    const testBundleId = await this.idb.installXCTestBundle(path.join(wdaBundlePath, 'PlugIns', 'WebDriverAgentRunner.xctest'));
+    return {wdaBundleId, testBundleId, wdaBundlePath};
+  }
+
+  async fetchWDABundle () {
+    if (!this.derivedDataPath) {
+      return await bundleWDASim();
```

Comment:
the name of this function is not very descriptive

Reply:
bundleWDASim or fetchWDABundle?

Reply:
the second one sounds better. it is always a good idea to have a verb in method names, although properties should be nouns

Reply:
bundle is also a verb